### PR TITLE
[dbwrappers] Optimize string-based field lookup

### PR DESCRIPTION
**Pull request by ksooo**
First three commits are to cleanup dbwrappers code a bit, no functional changes.

Last commit improves string-based field lookup, increasing overall performance for many use cases, among them creation of `CPVREpgInfoTag` instances with database data.

Before:
Screenshot 2022-08-24 at 09 03 09

After:
Screenshot 2022-08-24 at 09 03 40

### CODE REVIEW

File: xbmc/dbwrappers/dataset.h [hunk 1]
```diff
@@ -205,6 +205,7 @@ class Dataset
   Database* db; // info about db connection
   dsStates ds_state; // current state
   Fields *fields_object, *edit_object;
+  std::unordered_map<std::string, unsigned int> name2indexMap;
```
Comment (garbear) [inline xbmc/dbwrappers/dataset.h:208]: I like to annotate the key and value for better clarity, for example something like:

```c++
std::unordered_map<std::string, unsigned int> name2indexMap; // Lower case field name -> database index
```

File: xbmc/dbwrappers/dataset.cpp [hunk 2]
```diff
@@ -627,12 +576,22 @@ const char* Dataset::fieldName(int n)
     return NULL;
 }
 
+char* Dataset::str_toLower(char* s)
+{
+  for (char* p = s; *p; p++)
```
Comment (garbear) [inline xbmc/dbwrappers/dataset.cpp:581]: c++ version of for loop:

```c++
char* Dataset::str_toLower(char* s, size_t len)
{
  std::transform(s, s + len, s, [](unsigned char c) { return std::tolower(c); });
  return s;
}
```

**Comment by garbear** [comment]
Great PR. First the formatting, then the improvements (which are all red, nice and safe), then the change. The only debate we should see is on the "how" of the optimization.

File: xbmc/dbwrappers/dataset.cpp (hunk 2 above)
Comment (ksooo) [reply xbmc/dbwrappers/dataset.cpp:581]: `std::transform(s, s + len, s, [](char c) { return std::tolower(c); });`

I bet the profiler will reveal that this is significant slower than the 'raw' loop. And I'm doing all this to improve performance...

File: xbmc/dbwrappers/dataset.h (hunk 1 above)
Comment (ksooo) [reply xbmc/dbwrappers/dataset.h:208]: Thanks, I will add the comment.

File: xbmc/dbwrappers/dataset.cpp (hunk 2 above)
Comment (neo1973) [reply xbmc/dbwrappers/dataset.cpp:581]: The compiler generates the same code, see [Godbolt](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCAAzNIADqgKhE4MHt6%2BekkpjgJBIeEsUTHxtpj2eQxCBEzEBBk%2BflzllWk1dQQFYZHRcdIKtfWNWS2Dnd1FJf0AlLaoXsTI7BzmscHI3lgA1CaxblEMyAgsdQDWAPSHx6fEZwB0CHvYJhoAgmsbW5i7%2B2LAJEIJ2erw%2BZnWR2%2BvzcyAcAE8EpgQe9PpCvDs9m5BsRgsBkR93sc6gAqbbYgD6RACqAA7tEuBAicRSQpSGS8AAvTCU7b0BgzUEmADsVne2zJBHQIBABGIhgU/GILAgrLJu0svMYbNVJgArFZdQARRkIOrbZAzXYi7bETAERYMCVSmWeWnRRkC2IWK2Gz2it7i2324iOhR7f3Cw2Cwmm5kS4iU1DUunEMwmklkgUokWg8WK7bpuMJX6Gsnh7bEhLlquWaxZgPi8WVktO6VEWhu4gQSt%2B6MNoMOstewVCqMo96DZ3Y3HxxPJ6KxFWS6XToyZkf%2B8WTttyhgKkjKhT3KLAYIQOZk%2B6uc/a4%2BYU8MG%2B7fV641M82W4Xegch1sujsph65aRr2Yo2nag5hsO2ZjgSbwAG6oHg6DmgICg%2BNy2JLlOsozgo9Zfn2kZEeCXzoj8mLXCc5xXIwNznI8%2BKguMjjINsiHIdsFgALLkgyVG3Gc0odAQSJmAAbBKTCiZ%2BOZgfmEBMF4RDbOS2wgFJMlWpujYXBc2wACoIHgCioTswB2qZjhsOguaNtuICrsAap7KW5hmEZJnbF5LBwtsomDPGuL3O54Z2eKaB7hh5JYRSVKdgyR7oNJTBPkeKRcueMygQ2xEwaCenbAASveJmicQ/kID8VBeEcVTbEwplMNsAnnKCFjYKEbgABLcW8RUANIQDxfE5cxtSsexSEoSNaatXcwkTWJkksZgsk6QpSkqWpGmret4XbIVnmmWg5mWf5eA2QdDlOS5sRuWYHnGaZPl%2BQFBBBUYIWPWFYERWh0WxbKc6dmmSUpVlOXinlHyjgV%2BklaegzRJV1W1Q4aQNU1LV0dRdztZ1PV9YNw28WYY0ThNeBsRxM28Yu81CSAInLZpa3aXZm3KagqnqWz%2B1/Yd%2BnHWZPwWQQVmXZgtmCzduFrqZrnqk9Xmvf5mCBU531mL9Db/VFbAxbKS4JvFKaLg5LCoPBmAqtlUM%2BiOsGFYjZUowQVXbDVdWY41DU40ceNnATXW9f1Q0jbEOUcHMtCcLqvB%2BBwWikKgnBuNY1hkgsSwUeCPCkAQmgx3MQm6ho%2BicJIifF6nnC8AoIAV0Xycx6QcCwEgaAsAkdDROQlDd739AxJshjAFwsQaBXNC0OVjcQBEtcRMEdRwpwBfd2wggAPIMLQ6%2Bt6QWCnEY4hH/gtoYzbjdH5gqiYMgykrAXwSiXHR%2B0HgERysQcIeFgWuuEWAbzblQAwwAFAADU8CYBpDvREScC78EECIMQ7ApAyEEIoFQ6gj66BaAYIwKBM6WH0N/RusBmBsBAIzUgNsYjcEkFobKqcEhVFvgAWknK5UwtZLBcA5NsThO9YgNwqI/KoLgGDuE8E0fwMjJi9BiC0HIqQBAjGaKQNRVQlHFD6GMCRGMBAdGGHI0YrRJHtCGF0YIPR9EqNsDYzRehxj1D0dMLgcwFA52WHoWUmAX5tw/gnUgScU5pw4KoAAHOJTh4lJDmiIc5Se9wNBpILBnfhNhti4EICQdUsQWjbA8D3PuFU1heN4C3FhpcQDl0rhwauYTa6RIbk3QuxdWEfzMDXI%2BbTOmt1YQwlIzhJBAA%3D%3D). If you want to push performance make the function accept a `std::string` and also return one, this way move semantic can save an allocation. All three variants are in the link, you can simply open it in Quick-bench to see the difference (it isn't huge).

In general, the stuff from `<algorithm>` is as fast as hand rolled loops if they do the same.

Comment (ksooo) [reply xbmc/dbwrappers/dataset.cpp:581]: There is no “the compiler”.  I profiled  this in macOS and the std::transform version is definitely slightly slower, compared to the hand-crafted loop. I think I will keep the code as it is now. I’m not fighting for a couple of milliseconds.

Comment (howie-f) [reply xbmc/dbwrappers/dataset.cpp:581]: what does condition `*p` mean, same as `*p != NULL` or `*p != ‘\0‘`…?
definitely not `*p != nullptr` 🤔

Comment (ksooo) [reply xbmc/dbwrappers/dataset.cpp:581]: It checks the string for the 0 terminator.

**Pull request merged**